contracts/Strategy.sol

```diff
@@ -274,60 +371,317 @@ contract Strategy is BaseStrategy {
         (_liquidatedAmount, ) = liquidatePosition(type(uint256).max);
     }
 
-    // The _newStrategy must support the same kashiPair or bad things will happen
-    function prepareMigration(address _newStrategy) internal override {
-        kashiPair.transfer(_newStrategy, kashiFraction());
-    }
+    // We stake in the masterChef so we can't transfer assets when migrating
+    function prepareMigration(address _newStrategy) internal override {}
 
-    function setKashiPair(address _newKashiPair) external onlyGovernance {
-        require(
-            address(IKashiPair(_newKashiPair).bentoBox()) == address(bentoBox),
-            "BentoBox does not match"
-        );
+    function addKashiPair(address _newKashiPair, uint256 _newPid)
+        external
+        onlyGovernance
+    {
+        // cannot exceed max pair length
+        require(kashiPairs.length < MAX_PAIRS);
+        // must use the correct bentobox
         require(
-            IKashiPair(_newKashiPair).asset() == BIERC20(address(want)),
-            "KashiPair asset does not match want"
+            address(IKashiPair(_newKashiPair).bentoBox()) == address(bentoBox)
         );
+        // kashPair asset must match want
+        require(IKashiPair(_newKashiPair).asset() == BIERC20(address(want)));
+        if (_newPid != 0) {
+            // masterChef pid token must match the kashiPair
+            require(
+                address(masterChef.poolInfo(_newPid).lpToken) == _newKashiPair
+            );
+        }
+
+        for (uint256 i = 0; i < kashiPairs.length; i++) {
+            if (_newKashiPair == address(kashiPairs[i].kashiPair)) {
+                revert(); // kashiPair already attached
+            }
+        }
 
-        uint256 kashiFraction = kashiFraction();
+        kashiPairs.push(KashiPairInfo(IKashiPair(_newKashiPair), _newPid));
 
-        if (kashiFraction > 0) {
-            kashiPair.removeAsset(address(this), kashiFraction);
+        if (_newPid != 0) {
+            IERC20(_newKashiPair).safeApprove(
+                address(masterChef),
+                type(uint256).max
+            );
         }
+    }
 
-        kashiPair = IKashiPair(_newKashiPair);
+    function removeKashiPair(address _remKashiPair) external onlyGovernance {
```

Comment:
this function has not rug potential and might be needed to be called in an emergency-ish situation. 

maybe a onlyEmergencyAuthorized modifier is better?